Add new time format for parsing TAI seconds since 1958-01-01

wtbarnes opened the pull request:
<!--
We know that working on code and submitting pull requests takes effort, and we appreciate your time.
Thank you.

Please be aware that everyone has to follow our code of conduct:
https://sunpy.org/coc

Furthermore, you might need to check with your work place if you are allowed to contribute code:
https://docs.sunpy.org/en/latest/dev_guide/contents/newcomers.html

Also these comments are hidden when you submit this github pull request.

We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry about them!
Here is a brief explanation of them:
https://docs.sunpy.org/en/latest/dev_guide/contents/pr_review_procedure.html#continuous-integration
-->

### Description

<!--
Provide a general description of what your pull request does.

If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number GitHub will automatically link it.
If it doesn't, please remove the following line.
-->

This PR adds a new `astropy.time.Time` format for parsing the number of TAI (leap) seconds since 1958-01-01 00:00:00. This  was originally used in the SOHO CDS timing software (from what I can tell, see https://hesperia.gsfc.nasa.gov/ssw/gen/idl/time/aaareadme.txt) and is the format returned by the [`anytim2tai` SSW routine](https://hesperia.gsfc.nasa.gov/ssw/gen/idl/time/anytim2tai.pro).

I should say I am not dead set on the name for this time format. 'tai_cds' is a bit awkward and cryptic.

Comparisons between `astropy.time.Time` and `anytim2tai` show that after 1972-01-01, the difference between UTC and TAI is the same for both methods. However, between 1958-01-01 and 1972-01-01, the difference as returned  by `anytim2tai` is 9 s (jumping to 10 at 1972-01-01) while for `astropy.time.Time`, the difference increases roughly linearlly from 0 to 10. While this difference is not critical for missions after 1972, this should still be documented on our "Known Issues" page (see #5374).

![image](https://user-images.githubusercontent.com/8676141/128551211-bb0d5883-9941-4160-912c-20cec6655264.png)

dstansby reviewed (changes requested):
The oldestdeps build failure looks like it's caused by this PR.

wtbarnes commented:
Yeah, it looks like the `erfa` package was not an explicit dependency of astropy then?

wtbarnes commented:
So prior to astropy 4.2, PyERFA was included in astropy as a private module `_erfa`: https://docs.astropy.org/en/stable/whatsnew/4.2.html#removed-erfa-module.

wtbarnes commented:
I'm not sure why my import on the test side wasn't just the same as my import in the actual code for the time format class. I guess the question is, does 

```python
from astropy.coordinates.sky_coordinate import erfa
```

import the same thing as 

```python
from astropy.time.formats import erfa
```

Presumably the answer is yes? They are both going through astropy.

ayshih commented:
I'd like significant changes to the names and documentation:

- I propose the scale be named `tai_seconds` or something similar.
- The documentation should not state that CDS originated this scale.  Even if that's true – and I'm not sure that it is – it should be relegated to a note rather than as part of the name or as motivation.
- The documentation should not discuss `anytim2tai.pro` as if it were a reference implementation.  The differences with `anytim2tai.pro` should either be moved to docstring notes or to Known Issues.  `anytim2tai.pro` is flawed in two ways for pre-1972 times:
  - UT and TAI were synchronized by definition on 1958 Jan 1 00:00:00.  `anytim2tai.pro` should return 0 at that time.
  - Even if `anytim2tai.pro` is intentionally choosing ignoring the difference in lengths of UT seconds and TAI seconds prior to 1972, it should be returning 10 for pre-1972 times, not 9.  Its implementation acts as if there was a leap second added on 1972 Jan 1, but there was not.

wtbarnes commented:
> * I propose the scale be named `tai_seconds` or something similar.

Done. I've now renamed it to `TimeTAISeconds` with the format string `'tai_seconds'`

> * The documentation should not state that CDS originated this scale.  Even if that's true – and I'm not sure that it is – it should be relegated to a note rather than as part of the name or as motivation.

Fair. I've moved this to a note

> * The documentation should not discuss `anytim2tai.pro` as if it were a reference implementation.  The differences with `anytim2tai.pro` should either be moved to docstring notes or to Known Issues.  `anytim2tai.pro` is flawed in two ways for pre-1972 times:
>   
>   * UT and TAI were synchronized by definition on 1958 Jan 1 00:00:00.  `anytim2tai.pro` should return 0 at that time.
>   * Even if `anytim2tai.pro` is intentionally choosing to ignore the difference in lengths of UT seconds and TAI seconds prior to 1972, it should be returning 10 for pre-1972 times, not 9.  Its implementation acts as if there was a leap second added on 1972 Jan 1, but there was not.

Fair point. I did not intend to make it sound like a reference implementation, but I see how it does. I've shortened this to a note and will open a PR to add this to the known issues page.

samaloney commented on changelog/5489.feature.rst:
```diff
@@ -0,0 +1,4 @@
+Add ``sunpy.time.TimeTAICDS`` to support parsing the TAI time format originally used in the CDS
```
```suggestion
Add ``sunpy.time.TimeTAISeconds`` to support parsing the TAI time format originally used in the CDS
```

wtbarnes commented:
I've now called this `TimeUnixTai58` since it is exactly equivalent to `astropy.time.TimeUnixTai`, except that the epoch is 1958-01-01 instead of 1970-01-01. The name is kind of ugly, but is consistent with the astropy convention.

dstansby commented on sunpy/time/timeformats.py:
```diff
@@ -0,0 +1,75 @@
+    epoch is 12 years earlier.
+
+    .. note:: For dates on and after 1972-01-01 00:00:00 UTC, this format is
+              equivalent to that returned by the `anytim2tai` routine in SSW.
```
```suggestion
              equivalent to that returned by the ``anytim2tai`` routine in SSW.
```

dstansby commented on sunpy/time/timeformats.py:
```diff
@@ -0,0 +1,75 @@
+    .. note:: For dates on and after 1972-01-01 00:00:00 UTC, this format is
+              equivalent to that returned by the `anytim2tai` routine in SSW.
+              For dates between 1958-01-01 00:00:00 UTC and 1971-12-31 23:59:59 UTC,
+              `anytim2tai` returns a constant difference of 9 s difference between
```
```suggestion
              ``anytim2tai`` returns a constant difference of 9 s difference between
```

dstansby reviewed (commented):
Should fix the doc build

dstansby commented on sunpy/time/time.py:
```diff
@@ -13,7 +13,7 @@
 
 # This is not called but imported to register it
-from sunpy.time.utime import TimeUTime  # NOQA
+from sunpy.time.timeformats import TimeUTime  # NOQA
```
```suggestion
from sunpy.time.timeformats import *  # NOQA
```
Do we not want to register the new format too here?

wtbarnes replied:
Yes, but I'm a bit confused as to why this is needed here. I'm already importing it in `time/__init__.py` which should register it already?

dstansby replied:
👍 makes sense

wtbarnes replied:
Fixed

dstansby commented on sunpy/time/tests/test_unixtai58.py:
```diff
@@ -0,0 +1,50 @@
+from astropy.time.formats import erfa
+
+# This registers the TimeTAICDS format with astropy
+from sunpy.time import TimeUnixTai58  # NOQA
```
```suggestion
import sunpy.time  # NOQA
```
Should work too, and is a good test that we're importing TimeUnixTai58 into the `sunpy.time` namespace? This isn't a blocking suggestion though.

dstansby commented on sunpy/time/tests/test_unixtai58.py:
```diff
@@ -0,0 +1,50 @@
+    assert t2.unix_tai_58 - t1.unix_tai_58 == 10.0
+
+
+@pytest.mark.parametrize('time', [
```
This would be a good place to use `hypothesis` to generate test data. Not a blocking suggestion though.

ayshih reviewed (changes requested):
The name must be changed.  "Unix" should not be used here.  `TimeUnixTai` is named that way because Unix time is TAI seconds from 1970.  `TimeUnixTai1958` is self-contradicting.

ayshih commented on sunpy/time/tests/test_unixtai58.py:
```diff
@@ -0,0 +1,50 @@
+])
+def test_unix_tai_consistency(time):
+    diff_unix_tai = Time(TimeUnixTai.epoch_val, scale='tai', format=TimeUnixTai.epoch_format,).tai_seconds
+    assert time.unix_tai + diff_unix_tai == time.tai_seconds
```
```suggestion
```
This test appears to have negligible added value, because you're basically checking that `TimeFromEpoch` really means time from an epoch.  Essentially the only conceivable way that it could fail is if someone accidentally changes the `epoch_scale` of `TimeUnixTai` from TAI to something else, and that would presumably cause test failures in Astropy.

ayshih reviewed (commented):
Thanks for merging in wtbarnes/sunpy#2!

I've exchanged emails with Bill Thompson and Dominic Zarro, and Bill plans to add disclaimers to the SSW routines to make clear that they do not give valid results for pre-1972 dates.  For example, ``utc2tai.pro`` already has such a disclaimer.

ayshih commented on sunpy/time/tests/test_taiseconds.py:
```diff
@@ -0,0 +1,50 @@
+def test_roundtrip(time):
+    time_from_tai_seconds = Time(time.tai_seconds, scale='tai', format='tai_seconds')
+    assert time_from_tai_seconds == time
+
+
+@pytest.mark.parametrize('time', [
+    Time('1958-01-01T00:00:00', format='isot', scale='tai'),
+    Time('1972-01-01T00:00:00', format='isot', scale='tai'),
+    Time('2015-10-25T05:24:08', format='isot', scale='tai'),
+    Time('2018-09-17T19:46:25', format='isot', scale='tai'),
+])
+def test_unix_tai_consistency(time):
+    diff_unix_tai = Time(TimeUnixTai.epoch_val, scale='tai', format=TimeUnixTai.epoch_format,).tai_seconds
+    assert time.unix_tai + diff_unix_tai == time.tai_seconds
```
```suggestion
```
This test appears to have negligible added value, because you're basically checking that TimeFromEpoch really means time from an epoch. Essentially the only conceivable way that it could fail is if someone accidentally changes the epoch_scale of TimeUnixTai from TAI to something else, and that would presumably cause test failures in Astropy.

wtbarnes replied:
Fair enough. I've removed this.

nabobalis reviewed (changes requested):
Should have a what new section added before merge.

dstansby commented on docs/whatsnew/3.1.rst:
```diff
@@ -64,5 +64,19 @@ to use `~sunpy.map.GenericMap.date_average`,
 
+Addition of new time format `~sunpy.time.formats.TimeTaiSeconds`
+----------------------------------------------------------------
+The `~sunpy.time.formats.TimeTaiSeconds` format is the number of
```
```suggestion
The new `~sunpy.time.formats.TimeTaiSeconds` format is the number of
```

nabobalis commented:
Can we get a rebase to check the CI but the docs failed on:

```
/home/vsts/work/1/s/docs/whatsnew/3.1.rst:67: WARNING: py:obj reference target not found: sunpy.time.formats.TimeTaiSeconds
/home/vsts/work/1/s/docs/whatsnew/3.1.rst:69: WARNING: py:obj reference target not found: sunpy.time.formats.TimeTaiSeconds
```

dstansby commented on docs/whatsnew/3.1.rst:
```diff
@@ -64,5 +64,19 @@ to use `~sunpy.map.GenericMap.date_average`,
 `~sunpy.map.GenericMap.date_start`, or `~sunpy.map.GenericMap.date_end`
 instead if you need one of these specific times.
 
+Addition of new time format `~sunpy.time.formats.TimeTaiSeconds`
+----------------------------------------------------------------
```
```suggestion
Addition of new time format `~sunpy.time.TimeTaiSeconds`
--------------------------------------------------------
```

dstansby commented on docs/whatsnew/3.1.rst:
```diff
@@ -64,5 +64,19 @@ to use `~sunpy.map.GenericMap.date_average`,
 
+Addition of new time format `~sunpy.time.formats.TimeTaiSeconds`
+----------------------------------------------------------------
+The new `~sunpy.time.formats.TimeTaiSeconds` format is the number of
```
```suggestion
The new `~sunpy.time.TimeTaiSeconds` format is the number of
```

dstansby reviewed (commented):
Should fix the doc build
